src/com/amazon/corretto/crypto/provider/Utils.java

```diff
@@ -97,75 +97,77 @@ static int getMdLen(final long mdPtr) {
    * or if we're unable to determine whether they overlap. Unfortunately it's not possible to
-   * determine whether certain bytebuffers overlap (notably, if one buffer is a RO buffer and the
-   * other is an array-backed buffer, we cannot check for overlap without empirically modifying the
-   * array-backed buffer).
+   * determine whether certain bytebuffers overlap (notably, if one buffer is inputBuffer RO buffer
+   * and the other is an array-backed buffer, we cannot check for overlap without empirically
+   * modifying the array-backed buffer).
    *
    * <p>Overlap is determined based on buffer position and limit.
    */
-  static boolean buffersMaybeOverlap(ByteBuffer a, ByteBuffer b) {
-    boolean directA = a.isDirect();
-    boolean directB = b.isDirect();
-    boolean arrayA = a.hasArray();
-    boolean arrayB = b.hasArray();
+  static boolean outputClobbersInput(ByteBuffer inputBuffer, ByteBuffer outputBuffer) {
+    boolean inputIsDirect = inputBuffer.isDirect();
+    boolean outputIsDirect = outputBuffer.isDirect();
+    boolean inputHasArray = inputBuffer.hasArray();
+    boolean outputHasArray = outputBuffer.hasArray();
 
-    if ((directA || directB) && (directA != directB)) {
+    if ((inputIsDirect || outputIsDirect) && (inputIsDirect != outputIsDirect)) {
       // One is direct and the other isn't; there can be no overlap
       return false;
     }
 
-    if (directA && directB) {
+    if (inputIsDirect && outputIsDirect) {
       // By slicing the buffers, we can avoid having to think about the native pointer and
-      // position(); the
-      // position will simply be added into the native pointer.
+      // position(); the position will simply be added into the native pointer.
 
       // This will also allow getNativeBufferOffset to fully determine whether the buffers overlap
-      // in native code,
-      // by factoring the limit() into the buffer capacity.
-      return getNativeBufferOffset(a.slice(), b.slice()) <= Integer.MAX_VALUE;
+      // in native code, by factoring the limit() into the buffer capacity.
+      return getNativeBufferOffset(inputBuffer.slice(), outputBuffer.slice()) <= Integer.MAX_VALUE;
     }
 
     // At this point we'll need to check array() and arrayOffset(), but to do this we need both to
     // hasArray().
-    if (!(arrayA && arrayB)) {
+    if (!(inputHasArray && outputHasArray)) {
       // One doesn't hasArray, so we're prevented from checking for overlap. Return true to assume
       // overlap.
       return true;
     }
 
-    if (a.array() != b.array()) {
-      // different arrays, so no overlap
-      return false;
-    }
-
-    // Same array, check for overlap within array
-    long a_offset = (long) a.arrayOffset() + (long) a.position();
-    long b_offset = (long) b.arrayOffset() + (long) b.position();
-
-    if (a_offset > b_offset) {
-      return b_offset + (long) b.limit() > a_offset;
```

Comment:
This was the problematic logic which would trigger an allocation for all cases of the output being beyond the start of the input.